src/structures/interfaces/Collector.js

```diff
@@ -147,6 +147,38 @@ class Collector extends EventEmitter {
     });
   }
 
+  /**
+   * Allows this collector to be consumed with `for await...of`
+   * @type {AsyncIterator}
+   */
+  async * [Symbol.asyncIterator]() {
+    const queue = [];
+    const onCollect = item => {
+      queue.push(item);
+    };
+
+    this.on('collect', onCollect);
+
+    while (!this.ended || queue.length > 0) {
+      if (queue.length > 0) {
+        yield queue.shift();
+      } else {
+        await new Promise(resolve => { // eslint-disable-line no-await-in-loop
```

Comment:
There’s a subtle race condition here where an event might fire before the promise is created. You want to define the promise first so that event listeners are attached synchronously and only then await it i.e.
```suggestion
        const promise = new Promise(resolve => {
          const tick = () => {
            this.removeListener('collect', tick);
            this.removeListener('end', tick);
            resolve();
          };
          this.on('collect', tick);
          this.on('end', tick);
        });
        await promise;
```

Reply:
That's really not the concern of this nice to have feature. And would only occur if something happens async before creating the structure and using it as an async iterator, which pretty well shouldn't be the case, and would have the same issue if they were using it as normal when attaching an event listener.

Reply:
> And would only occur if something happens async before creating the structure and using it as an async iterator

You’re definitely more familiar with how this would be used, I’m just saying that `await` pauses execution, and the promise might miss a `collect` or `end` event in the intervening time, resulting in the promise here hanging. Again, just a suggestion 🙂.

Reply:
That's not correct. the callback from a promise constructor is called sync immediately, so the event handlers won't miss anything.

```js
console.log(1)
const myPromise = new Promise(resolve => {
  console.log(2);
  resolve();
});

console.log(3);

myPromise.then(() => console.log(5));

console.log(4);
```
As proof, you can run that, and you will see it logs 1-5 in order.

Reply:
I was worried about the `await` not the Promise executor, but I just tested it and it seems like the expression that is awaited is run synchronously. My bad.
```
(async () => {
console.log(1);
Promise.resolve().then(() => console.log(4));
new Promise(resolve => {
  console.log(2);
  resolve();
});
await new Promise(resolve => {
  console.log(3);
  resolve();
});
})();
```

Reply:
Alright, now that we are on the same page, can you mark this change as resolved so it doesn't confuse the pr author?

Reply:
No button in the interface to do this, I think that’s if I clicked “request changes” in the review but I didn’t